src/main/java/QueryServlet.java

```diff
@@ -218,12 +244,134 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+              ImmutableList.of(add, add, percent))
+          .put(
+              "P014001,P014021,P014022,P014042,P014043",
+              ImmutableList.of(rightSubtract, rightSubtract, rightSubtract, rightSubtract))
+          .build();
+
+  private String reformatDataArray(String data, String dataIdentifier, boolean isCountyQuery)
+      throws InvalidObjectException {
+    if (!dataRowToReformatFunction.containsKey(dataIdentifier)) {
+      return data; // This data doesn't need reformatting
+    }
+    if ((dataRowToReformatFunction.get(dataIdentifier).size() + 1)
+        != dataIdentifier.split(",").length) {
+      // We should have enough functions listed to apply them to each group
+      // of two rows - (a, b) and then (their product, c), etc.
+      throw new InvalidObjectException(
+          "There aren't the right number of functions to apply to these data columns.");
+    }
+
+    // Convert from the census API's JSON string to a Java matrix
+    Gson gson = new Gson();
+    Type dataArrayType = new TypeToken<ArrayList<List<String>>>() {}.getType();
+    ArrayList<List<String>> originalDataArray = gson.fromJson(data, dataArrayType);
+    ArrayList<List<String>> newDataArray = new ArrayList<List<String>>();
+
+    // Create a new matrix to contain the reformatted data. Add in the the first column, which is
+    // the string identifier of the location, and the second column, the first numerical value
+    for (int i = 0; i < originalDataArray.size(); i++) {
+      List<String> originalDataRow = originalDataArray.get(i);
+      List<String> newDataRow = new ArrayList<String>();
+      newDataRow.add(originalDataRow.get(0));
+      newDataRow.add(originalDataRow.get(1));
+      newDataArray.add(newDataRow);
+    }
+
+    // Combine the original numbers, skipping the header row, by iterating over the functions
+    // given for these data columns. For each row, we successively combine the next datapoint
+    // with the previous result, finally resulting in one single fully-combined data point per row.
+    List<BiFunction> numberCombiners = dataRowToReformatFunction.get(dataIdentifier);
+    for (int i = 0; i < numberCombiners.size(); i++) { // Moving across columns combining numbers
+      BiFunction numberCombiner = numberCombiners.get(i);
+      for (int j = 1; j < originalDataArray.size(); j++) { // Moving down each row
+        List<String> originalDataRow = originalDataArray.get(j);
+        List<String> newDataRow = newDataArray.get(j);
+        if (newDataRow.get(1) == null || newDataRow.get(1).startsWith("-")) {
+          // Sometimes the census API returns missing or negative numbers by mistake;
+          // these get cleaned up in the JavaScript
+          newDataRow.set(1, "-1");
+        } else {
+          try {
+            int newValue =
+                (int)
+                    numberCombiner.apply(
+                        Float.parseFloat(newDataRow.get(1)), // Always replacing previous value
```

Comment:
It seems to be that you are performing string to float and then float to string conversions too many times. That is probably because of the way in the which the two loops are structured. If you switch them around then you work through rows of data and at the beginning you can convert to float once, apply all the transformations as you work through entries in numberCombiners and then finally write back as string once.

Reply:
That's a good idea, thank you.

Reply:
Done.